File: Sep/Day16/replace-non-coprime-numbers-in-array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []

        for num in nums:
            while stack:
                gcd = math.gcd(stack[-1], num)
                if gcd == 1:
                    break
                num = (stack.pop() * num) // gcd # LCM = (num1 * num2) // gcd
            stack.append(num)

        return stack

File: Sep/Day16/replace-non-coprime-numbers-in-array.js
class Solution {
    replaceNonCoprimes(nums) {
        const stack = [];
        for (let num of nums) {
            while (stack.length > 0) {
                const g = this.gcd(stack[stack.length - 1], num);
                if (g === 1) break;
                num = Math.floor((stack.pop() * num) / g); // LCM = (a * b) / gcd
            }
            stack.push(num);
        }
        return stack;
    }

    gcd(a, b) {
        while (b !== 0) {
            let temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}
